dbms/src/Interpreters/Join.cpp

```diff
@@ -2841,21 +2841,31 @@ void Join::waitUntilAllProbeFinished() const
 
 void Join::finishOneNonJoin(size_t partition_index)
 {
-    while (partition_index < build_concurrency)
+    if likely (active_build_threads == 0 && active_probe_threads == 0)
     {
-        std::unique_lock partition_lock = partitions[partition_index]->lockPartition();
-        partitions[partition_index]->releaseBuildPartitionBlocks(partition_lock);
-        partitions[partition_index]->releaseProbePartitionBlocks(partition_lock);
-        if (!partitions[partition_index]->isSpill())
+        /// only clear hash table if not active build/probe threads
+        while (partition_index < build_concurrency)
         {
-            releaseBuildPartitionHashTable(partition_index, partition_lock);
+            std::unique_lock partition_lock = partitions[partition_index]->lockPartition();
+            partitions[partition_index]->releaseBuildPartitionBlocks(partition_lock);
+            partitions[partition_index]->releaseProbePartitionBlocks(partition_lock);
+            if (!partitions[partition_index]->isSpill())
+            {
+                releaseBuildPartitionHashTable(partition_index, partition_lock);
+            }
+            partition_index += build_concurrency;
         }
-        partition_index += build_concurrency;
     }
 }
 
 Block Join::joinBlock(ProbeProcessInfo & probe_process_info) const
 {
+    if unlikely (active_build_threads != 0)
```

Comment:
Do we have other ways to check cancel situation, like isCancelledOrThrowIfKilled?
Others LGTM

Reply:
We can't use `isCancelledOrThrowIfKilled` here because when one thread meet eof, it will call `cancel()`, and in this case, other threads should not affected by that `cancel()`